synapse/storage/state.py

```diff
@@ -460,7 +511,7 @@ def _get_state_group_for_events(self, event_ids):
 
         defer.returnValue({row["event_id"]: row["state_group"] for row in rows})
 
-    def _get_some_state_from_cache(self, group, types):
+    def _get_some_state_from_cache(self, group, types, filtered_types=None):
```

Comment:
I'm not sure this method does the right thing with filtered_types.

Reply:
i've stared at this some more and am failing to guess or see why you think it's doing the wrong thing.

Reply:
ok, suppose we have a partial cache of the state, including the `m.room.member` events, but not the `m.room.history_visibility` events. Suppose this is called with:

```
store._get_some_state_from_cache(group, [('m.room.member', 'user1')], filtered_types= ['m.room.member'])
```

We have all the events we need to satisfy the `types` list, so `missing_types` will be empty (and hence `got_all` will be True), even though we are missing important state.

Reply:
right, missed that - thanks. hopefully e22700c3d resolves this.

Reply:
ok, so now what happens when we have

```
store._get_some_state_from_cache(group, [], filtered_types= ['m.room.member'])
```

?

At this point I feel like I'm being a human unit-test rig, and it makes me sad.

Reply:
sorry, i totally missed the messy [] special case. if it's any consolation, i strongly suspect i'd have missed it in a UT too.

Reply:
I still don't think this is right; I don't think e22700c actually fixed the case I mentioned.

Prove me wrong with a test!

> i strongly suspect i'd have missed it in a UT too.

Well, can you add one now, please.